### PR TITLE
Move 'deleted_utc' into 'state_user_deleted_utc' and change types.

### PR DESCRIPTION
**Pull request by zorbathut**

### CODE REVIEW

File: migrations/versions/2023_06_11_09_09_00_8337558f4f36_move_deleted_utc_into_state_user_.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+# revision identifiers, used by Alembic.
+revision = '8337558f4f36'
+down_revision = '6403c9151c12'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (justcool393) [inline migrations/versions/2023_06_11_09_09_00_8337558f4f36_move_deleted_utc_into_state_user_.py]: major issue with the revision here from what i can tell is that it undeletes every single post or comment in the database

ideally we should... not do that.

File: migrations/versions/2023_06_11_09_09_00_8337558f4f36_move_deleted_utc_into_state_user_.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    op.add_column('comments', sa.Column('state_user_deleted_utc', sa.DateTime(), nullable=True))
```
Comment (TLSM) [inline migrations/versions/2023_06_11_09_09_00_8337558f4f36_move_deleted_utc_into_state_user_.py]: Postgres has both `timestamp [without time zone]` and `timestamp with time zone`. These are the same size on disk. SQLA turns the former into a tz-naive datetime and the latter into a tz-aware datetime. The latter are often much easier to work with, in my experience, since Python's stdlib `datetime` treats them both differently (despite both being `datetime.datetime` objects) and sometimes requires tz-awareness (which you will only discover at runtime).

This migration appears to create `timestamp without time zone` columns. Would recommend finding a way to make them tz-aware.

File: migrations/versions/2023_06_11_09_09_00_8337558f4f36_move_deleted_utc_into_state_user_.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    op.drop_column('comments', 'deleted_utc')
```
Comment (TLSM) [inline migrations/versions/2023_06_11_09_09_00_8337558f4f36_move_deleted_utc_into_state_user_.py]: This will un-delete all currently deleted comments.

The SQL to do this and the previous comment correctly is something like:
```sql
ALTER TABLE comments ALTER COLUMN deleted_utc TYPE timestamp with time zone USING to_timestamp(deleted_utc);
ALTER TABLE comments RENAME COLUMN deleted_utc TO state_user_deleted_utc;
```

File: files/routes/comments.py [hunk 4]
```diff
@@ -332,9 +332,9 @@ def edit_comment(cid, v):
 @auth_required
 def delete_comment(cid, v):
 	c = get_comment(cid, v=v)
-	if c.deleted_utc: abort(409)
+	if c.state_user_deleted_utc: abort(409)
 	if c.author_id != v.id: abort(403)
-	c.deleted_utc = int(time.time())
+	c.state_user_deleted_utc = time.time()
```
Comment (TLSM) [inline files/routes/comments.py]: This still tries to set it to a unix timestamp (and causes a stack trace in testing).
Recommend using `datetime.now(tz=timezone.utc)` (see later comments on tz-aware datetimes).
Necessary imports if not already available: `from datetime import datetime, timezone`

File: files/routes/posts.py [hunk 5]
```diff
@@ -603,7 +603,7 @@ def delete_post_pid(pid, v):
 	if post.author_id != v.id:
 		abort(403)
 
-	post.deleted_utc = int(time.time())
+	post.state_user_deleted_utc = datetime.now()
```
Comment (TLSM) [inline files/routes/posts.py]: This works as written, but see other comments on tz-aware datetimes.
Recommend using `datetime.now(tz=timezone.utc)`
Necessary imports if not already available: `from datetime import datetime, timezone`

File: files/templates/submission.html [hunk 6]
```diff
@@ -512,7 +512,7 @@ <h5 class="card-title">Jump in the discussion.</h5>
 	</script>
 {% endif %}
 
-{% elif not p.replies and p.deleted_utc == 0 %}
+{% elif not p.replies and p.state_user_deleted_utc == None %}
```
Comment (TLSM) [inline files/templates/submission.html]: I think it's actually canonically `none` in Jinja (though they might have aliased `None` in recent versions).
Jinja also has a template test for none-ness: `p.state_user_deleted_utc is none`

File: files/classes/cron/submission.py [hunk 7]
```diff
@@ -80,8 +80,8 @@ def make_title(self, trigger_time:datetime) -> str:
 	# HTML template for previewing a submitted task
 
 	@property
-	def deleted_utc(self) -> int:
-		return int(not self.task.enabled)
+	def state_user_deleted_utc(self) -> int:
```
Comment (TLSM) [inline files/classes/cron/submission.py]:
```py
def state_user_deleted_utc(self) -> datetime | None:
    return datetime.now(tz=timezone.utc) if not self.task.enabled else None
```

**Review by TLSM — CHANGES_REQUESTED** [review]
Main points:
1) Current migration un-deletes all currently deleted content.
2) While fixing that, recommend switching to tz-aware datetimes in Python and `timestamp with time zone` columns in Postgres.
3) Comment deletion currently still uses timestamps. Recommend switching both it and post deletion to use tz-aware datetimes.

**Review by justcool393 — DISMISSED** [review]
see @TLSM's comments

File: migrations/versions/2023_06_11_09_09_00_8337558f4f36_move_deleted_utc_into_state_user_.py (hunk 2 above)
Comment (zorbathut) [reply migrations/versions/2023_06_11_09_09_00_8337558f4f36_move_deleted_utc_into_state_user_.py]: Sure, done!

File: migrations/versions/2023_06_11_09_09_00_8337558f4f36_move_deleted_utc_into_state_user_.py (hunk 3 above)
Comment (zorbathut) [reply migrations/versions/2023_06_11_09_09_00_8337558f4f36_move_deleted_utc_into_state_user_.py]: Dammit, could've sworn I'd written update code. I think it got accidentally nuked when I was tinkering with the changelist >_<

Anyway, fixed.

File: migrations/versions/2023_06_11_09_09_00_8337558f4f36_move_deleted_utc_into_state_user_.py (hunk 1 above)
Comment (zorbathut) [reply migrations/versions/2023_06_11_09_09_00_8337558f4f36_move_deleted_utc_into_state_user_.py]: Dammit, could've sworn I'd written update code. I think it got accidentally nuked when I was tinkering with the changelist >_<

Anyway, fixed.

File: files/routes/comments.py (hunk 4 above)
Comment (zorbathut) [reply files/routes/comments.py]: Should now be fixed.

File: files/routes/posts.py (hunk 5 above)
Comment (zorbathut) [reply files/routes/posts.py]: Should now be fixed.

File: files/templates/submission.html (hunk 6 above)
Comment (zorbathut) [reply files/templates/submission.html]: Should now be fixed.

File: files/classes/cron/submission.py (hunk 7 above)
Comment (zorbathut) [reply files/classes/cron/submission.py]: Should now be fixed.

**Review by TLSM — APPROVED** [review]
Briefly tested post & comment deletion/restoration through the UI, ensured the scheduled post logic uses its mocked deletion state correctly, and partially tested the DB migration. All appears to be working! Thanks for addressing review comments.

Perhaps wise to take a DB backup first: the migration looks correct to me and appears to work correctly, but Lord knows this application puts some awfully strange things into production databases.
